Scripts/WikiUpdater.cs

```diff
@@ -0,0 +1,379 @@
+
+            string text;
+            switch (child.TagName)
+            {
+                case "P":
+                    text = ConvertTextToBbcode(child.InnerHtml) + "\n\n";
+                    break;
+                case "UL":
+
+                    // Godot 3 does not support lists in BBCode, so use custom formatting
+                    text = child.Children
+                        .Where(c => c.TagName == "LI")
+                        .Select(li => $"[indent]•   {ConvertTextToBbcode(li.InnerHtml)}[/indent]")
+                        .Aggregate((a, b) => a + "\n" + b) + "\n\n";
+                    break;
+                default:
+                    // Ignore all other tag types
+                    continue;
+            }
+
+            // Concatenate this tag with the rest of the section so far
+            sections[^1] = new Wiki.Page.Section(sections[^1].SectionHeading, sections[^1].SectionBody + text);
+        }
+
+        return sections.Select(s => new Wiki.Page.Section(s.SectionHeading, s.SectionBody.Trim())).ToList();
+    }
+
+    /// <summary>
+    ///   Returns an equivalent section of a wiki page where the heading and body have been replaced with appropriate
+    ///   translation keys.
+    /// </summary>
+    private static Wiki.Page.Section UntranslateSection(Wiki.Page.Section section, string pageName)
+    {
+        var sectionName = section.SectionHeading?.ToUpperInvariant().Replace(" ", "_");
+        var heading = sectionName != null ? $"WIKI_HEADING_{sectionName}" : null;
+        var body = sectionName != null ? $"WIKI_{pageName}_{sectionName}" : $"WIKI_{pageName}_INTRO";
+
+        return new Wiki.Page.Section(heading, body);
+    }
+
+    /// <summary>
+    ///   Converts HTML for a single paragraph into BBCode. Paragraph must not contain lists, headings, etc.
+    /// </summary>
+    private static string ConvertTextToBbcode(string paragraph)
+    {
+        // Process our custom BBCode first
+        foreach (var compound in CustomBbcodeCompounds)
+        {
+            var compoundText = compound.ToLowerInvariant().Replace(" ", string.Empty);
+            paragraph = paragraph.Replace(
+                $"<b>{compound}</b>",
+                $"[thrive:compound type=\"{compoundText}\"][/thrive:compound]");
+        }
+
+        return paragraph
+            .Replace("\n", string.Empty)
+            .Replace("<b>", "[b]")
+            .Replace("</b>", "[/b]")
+            .Replace("<i>", "[i]")
+            .Replace("</i>", "[/i]")
+            .Replace("\"", "\\\"");
+    }
+
+    /// <summary>
+    ///   Inserts into en.po the English translations for all the translation keys in a list of wiki pages.
+    /// </summary>
+    private static async Task InsertTranslatedPageContent(List<TranslationPair> pages,
+        CancellationToken cancellationToken)
+    {
+        var lines = new List<string>(
+            await File.ReadAllLinesAsync(ENGLISH_TRANSLATION_FILE, Encoding.UTF8, cancellationToken));
```

Comment:
Well that would have been useful to have noticed...

Reply:
Or not. I keep hitting an impasse here.

That package seems to ignore obsolete values, so while I can just about retain everything else from en.po when I parse and regenerate, I lose the commented out strings at the end of the file.

The other option I found was msgcat, which should be able to combine .po files, but keeps messing up the output by reformatting file location comments. It keeps splitting them over multiple lines so breaks the localization update next time it runs. Maybe there is a way to get it to stop doing that, but [the options are limited](https://www.gnu.org/software/gettext/manual/html_node/msgcat-Invocation.html). The width option doesn't help.

Reply:
I guess it is then unavoidable to need to do a bit of manual handling here. It's at least not too bad to correctly detect the place where text content needs to be inserted into / replace existing lines in the file. Unless removing the old translations is fine, which I've been thinking about cleaning at some point but I suppose it isn't the best if this script always clears those out.

Reply:
Are you happy to leave this method as it is then? It's been resilient every time I've run the script.

Reply:
I think the overall approach is probably fine, but I might want to comment on some of the specifics. Like it looks like the script reads all of the file into memory, rather than doing an operation like reading line by line and maybe even outputting lines as it goes to another writer (and swapping then the file to the right location after writing the full file). Though I can't exactly remember off the top of my head if there was an async way to read a file line by line. If there isn't then this is probably fine. If I found a way to do that it probably is used somewhere in the file checks with the scripts / a base class in the common repo.

Reply:
I've rewritten the algorithm to asynchronously run through each line, removing or adding extra lines as needed. Tested on the versions of en.po from this branch and master, and it still produces the correct output in both cases.